Force send a mapblock to a player.

sofar opened the pull request:
Send a single mapblock to a specific remote player.

This is badly needed for mods and games where players are teleported
into terrain which may be not generated, loaded, or modified
significantly since the last player visit.

In all these cases, the player currently ends up in void, air, or
inside blocks which not only looks bad, but has the effect that the
player might end up falling and then the server needs to correct for
the player position again later, which is a hack.

The best solution is to send at least the single mapblock that the
player will be teleported to. I've tested this with ITB which does this
all the time, and I can see it functioning as expected (it even shows
a half loaded entry hallway, as the further blocks aren't loaded yet).

The parameter is a blockpos (table of x, y, z), not a regular pos.

The function returns a boolean indicating whether the call was successful

paramat commented:
> but has the effect that the player might end up falling

Generally, a player teleported to an unloaded or ungenerated area will not fall as the nodes surrounding the player will be "ignore". However, there may be special situations where falling does happen, and the feature is valid for other reasons.

sofar commented:
I was very careful to avoid going into the details but list all possible outcomes of teleporting a player of the client has an outdated perception of the mapblock.

On itb, players fall, 95% of the time, when teleported to a newly emerged box. It is not a bizarre weird bug, purely a MT design problem that this commit resolves. The benefits to other teleporting mods are self evident: a much more enjoyable teleport experience.

sofar commented:
- Added API documentation
- Fixed a locking bug if something failed
- Added return value to signal success

nerzhul commented on src/script/lua_api/l_object.cpp:
```diff
@@ -580,6 +580,29 @@ int ObjectRef::l_get_eye_offset(lua_State *L)
+
+	// player
+	RemotePlayer *player = getplayer(ref);
+	if (player == NULL)
```
if (!player)

nerzhul commented on src/server.cpp:
```diff
@@ -2313,6 +2313,25 @@ void Server::SendBlocks(float dtime)
+	if (!block)
+		return false;
+
+	m_clients.lock();
```
can we use std::lock_guard<> here instead of raw locking ?

sofar commented:
> if (!player)

Fixed

> can we use std::lock_guard<> here instead of raw locking ?

I can understand you asking this, but given that there are 0 uses of
this in the current codebase for this, I feel that it's not really
reasonable. Using this method should be done consistently with good
code examples and by people who understand the mechanism, and you're not
setting me up for success if you ask me to do this, because all of those
elements are missing now.

sofar commented:
I added an exception handler for getBlockNoCreate() as on testing
it was found to throw due to mapblocks being invalid. After that
it seems robust enough: I've thrown it at a player with random blocpos
values and it didn't budge.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -5240,6 +5240,8 @@ This is basically a reference to a C++ `ServerActiveObject`
     * in third person view (max. values `{x=-10/10,y=-10,15,z=-5/5}`)
 * `get_eye_offset()`: returns `offset_first` and `offset_third`
+* `send_mapblock(blockpos)`: returns `false` if failed. Sends a forced
+    mapblock to the remote player.
```
Not necessary a forced mapblock. `Sends a server-side loaded mapblock to the player.`

sofar replied:
Updated. Check the latest version.

SmallJoker commented:
Your changes looks good; just one comment.
What if we could automate this process so that the server always sends a yet (client-side) unloaded mapblock to the client on `set_pos`?

sofar commented:
It's a lot better with this patch, but still not 100%. Possibly due to reordering or just mapblock handling time in the client. ITB is also special in that it cleans up boxes with air all the time, so the risk of falling is high. For most games, this is much less likely.

Just from that perspective, I'd say don't bother for now.

paramat commented:
I disagree with automation. Often this won't be needed on player 'set pos', and it's more flexible not automated.

SmallJoker commented:
@paramat It improves the gameplay feeling when the target mapblock is known early by the client so that they can start walking right off after teleporting. Mapblocks known to the player can be skipped easily.
```C++
// Server::SendBlock
if (!client || client->isBlockSent(blockpos)) {
		m_clients.unlock();
		return false;
}
```
A call to `SendBlock` for `set_pos` for players would be quite cheap.

nerzhul commented:
i agree with @SmallJoker

paramat commented:
Hmm good point actually.
But, wouldn't the new mapblock be sent immediately anyway when 'set pos' happens? I can't see how it would get sent any earlier.

paramat commented:
Ok thanks, no objection from me then, can't think of a disadvantage :)

sofar commented:
It doesn't, set_pos happens immediately but I think the mapblocks are sent when the active block update interval hits, so there is a delay.

sofar commented:
@SmallJoker I looked into adding the set_pos but I can't seem to find a good place to add this? Would you add this in the `/script` side of things or somewhere in the engine `/client` code?

SmallJoker commented:
@sofar I'd personally extend `PlayerSAO::setPos` and `PlayerSAO::moveTo` with this code, so that mapblocks are sent instantly on respawn or by any other movement action (including `obj:set_pos()`).

DS-Minetest commented on doc/lua_api.txt:
```diff
@@ -5240,6 +5240,9 @@ This is basically a reference to a C++ `ServerActiveObject`
     * in first person view
     * in third person view (max. values `{x=-10/10,y=-10,15,z=-5/5}`)
 * `get_eye_offset()`: returns `offset_first` and `offset_third`
+* `send_mapblock(blockpos)`:
```
I can't find much documentation to `blockpos`. The only thing I found is:
https://github.com/minetest/minetest/blob/77961aa73d1947a2eabaa832319b7d73db6d13c2/doc/lua_api.txt#L4094-L4095
Which isn't really helpful.
How should you know where a block begins? (Eg. is there a block from x = 0 to x = 15 or x = -2 to x = 13 or whatever?)
It would be useful to have helper functions for this.

paramat replied:
'blockpos' needs clearly documenting and explaining certainly. Will make an issue to remind myself.

paramat replied:
Documented in #8376

SmallJoker replied:
`mapblock` would be less confusing here.

nerzhul commented on src/script/lua_api/l_object.cpp:
```diff
@@ -580,6 +580,29 @@ int ObjectRef::l_get_eye_offset(lua_State *L)
+	// pos
+	v3s16 p = read_v3s16(L, 2);
+
+	// Do it
```
useless com

DS-Minetest replied:
This comment is in all the other functions in this file, too.

sofar replied:
I'll nuke it though, it annoys me too :)

sofar commented:
- Fixed the little lua comment thing in the top commit
 - Added the requested mapblock send on `playerSAO::setPos()`

I think this is ready for merge now.

nerzhul commented on doc/lua_api.txt:
```diff
@@ -5324,6 +5324,9 @@ This is basically a reference to a C++ `ServerActiveObject`
     * in third person view (max. values `{x=-10/10,y=-10,15,z=-5/5}`)
 * `get_eye_offset()`: returns `offset_first` and `offset_third`
+* `send_mapblock(blockpos)`:
+    * Sends a server-side loaded mapblock to the player.
```
i suggest to add a warning that this is a heavy instruction and it should not be used too many

sofar replied:
Added an extra line here.

rubenwardy commented on src/content_sao.cpp:
```diff
@@ -1193,6 +1193,10 @@ void PlayerSAO::setPos(const v3f &pos)
 
+	// Send mapblock of target location
+	v3s16 blockpos = v3s16(pos.X / MAP_BLOCKSIZE, pos.Y / MAP_BLOCKSIZE, pos.Z / MAP_BLOCKSIZE);
+	(void) m_env->getGameDef()->SendBlock(m_peer_id, blockpos);
```
I guess this is to indicate that the result is discarded?

sofar replied:
Dropping it since it appears this code style isn't used anymore (it's technically outdated anyway)

sofar commented:
Sorry, yes, habits of C here of mine

sofar commented on src/server.cpp:
```diff
@@ -2313,6 +2313,28 @@ void Server::SendBlocks(float dtime)
+	if (!block)
+		return false;
+
+	m_clients.lock();
```
Can you please write more than 3 words here? I do not understand "it's not solved". Please explain how "it's not solved" so I can attempt to understand what that statement means.

nerzhul replied:
replace lock/unlock with std::lock_guard. As lock & unlock functions are defined the lock_guard can replace manual lock/unlock. Please look at other code parts using that

sofar replied:
```
$ git pull --rebase origin master
X11 forwarding request failed on channel 0
remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (29/29), done.
remote: Total 35 (delta 28), reused 29 (delta 28), pack-reused 6
Unpacking objects: 100% (35/35), done.
From github.com:minetest/minetest
 * branch              master     -> FETCH_HEAD
   e22a69d6..eadcbe47  master     -> origin/master
First, rewinding head to replay your work on top of it...
Applying: Force send a mapblock to a player.
Applying: Always send mapblock on teleport or respawn.
$ grep -ri guard src
src/debug.cpp:	case EXCEPTION_GUARD_PAGE:
src/debug.cpp:		return "Attempted guard page access";
$
```

sofar replied:
Invalid request. Please implement this first in a few places so that it becomes established code style and locking protection. After that, we can use it as a requirement for new commits to adhere to this newly established coding standard.

Until this is an established coding standard, the request is invalid.

rubenwardy replied:
Do we use some other form of RAII locks elsewhere?

nerzhul replied:
@rubenwardy yes we use them in some places

DS-Minetest replied:
As far as I can see, [lock_guard is not used anywhere in minetest](https://github.com/minetest/minetest/search?q=lock_guard&unscoped_q=lock_guard), but it's done like this:
https://github.com/minetest/minetest/blob/b298b0339c79db7f5b3873e73ff9ea0130f05a8a/src/util/container.h#L86

Maybe add a function
```cpp
MutexAutoLock& lock() { return MutexAutoLock(m_clients_mutex); }
```
to `ClientInterface`?

sofar commented:
There's one question open to @nerzhul, all the other comments were properly addressed, I think

SmallJoker commented on src/content_sao.cpp:
```diff
@@ -1193,6 +1193,10 @@ void PlayerSAO::setPos(const v3f &pos)
 		return;
 
+	// Send mapblock of target location
+	v3s16 blockpos = v3s16(pos.X / MAP_BLOCKSIZE, pos.Y / MAP_BLOCKSIZE, pos.Z / MAP_BLOCKSIZE);
```
`v3s16 blockpos =  floatToInt(pos, BS);`

sofar replied:
No! `BS` == 10. `MAP_BLOCKSIZE` == 16

sofar replied:
I don't think `floatToInt` is the correct conversion either here.

SmallJoker replied:
Oops, you're totally right.

SmallJoker commented on src/script/lua_api/l_object.cpp:
```diff
@@ -583,6 +583,25 @@ int ObjectRef::l_get_eye_offset(lua_State *L)
+	RemotePlayer *player = getplayer(ref);
+	if (!player)
+		return 0;
+	v3s16 p = read_v3s16(L, 2);
```
`v3s16 p =check_v3s16(L, 2);` superfluous L591.

SmallJoker commented on src/server.cpp:
```diff
@@ -2314,6 +2314,28 @@ void Server::SendBlocks(float dtime)
+
+	m_clients.lock();
+	RemoteClient *client = m_clients.lockedGetClientNoEx(peer_id, CS_Active);
+	if (!client) {
```
`|| client->isBlockSent(block)`
I don't think it makes sense to resend already known blocks.

sofar replied:
It needed to be `blockpos` but sure!

sofar commented:
The open question was invalid. I see no further comments or requested changes at this time.

sofar commented:
Remaining comments / requests are invalid, as far as I can see.

sofar commented:
Can I get someone to send some :+1:'s now, please?

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -5380,6 +5380,10 @@ This is basically a reference to a C++ `ServerActiveObject`
+* `send_mapblock(blockpos)`:
+    * Sends a server-side loaded mapblock to the player.
+    * Returns `false` if failed.
+    * Resource intensive - use sparsely
```
should say something like: to get the mapblock, simply divide the position by 16.
The engine hopefully would round automagically

sofar replied:
geeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeesh

sofar replied:
fixed with an exact and better description.

paramat replied:
I've actually added an explanation of getting nodepos from blockpos and blockpos from nodepos, elsewhere in lua_api.txt so no need to explain all that again.

rubenwardy commented on src/content_sao.cpp:
```diff
@@ -1193,6 +1193,10 @@ void PlayerSAO::setPos(const v3f &pos)
 
+	// Send mapblock of target location
+	v3s16 blockpos = v3s16(pos.X / MAP_BLOCKSIZE, pos.Y / MAP_BLOCKSIZE, pos.Z / MAP_BLOCKSIZE);
+	m_env->getGameDef()->SendBlock(m_peer_id, blockpos);
```
How often is setPos called? Ie: is it called when the client sends a position update

sofar replied:
No, only when the server, though lua, sends a location change (teleport)

sofar replied:
This includes /home and resurrection.

rubenwardy commented:
I've seen this used in code:

```c++
MutexAutoLock lock(m_spawner_list_lock);
```

will need to look at what `m_clients.lock()` is

sofar commented:
> I've seen this used in code:

That shouldn't prevent this from getting merged.

sofar commented:
>@rubenwardy yes we use them in some places

#define we
#define them
#define some_places

paramat commented:
@SmallJoker are you able to approve? This looks close to being approved.
